File: 0056_mergeIntervals/main.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        intervals.sort(key = lambda i : i[0])
        prev = intervals[0]
        for interval in intervals[1:]:
            if prev[1] < interval[0]:
                res.append(prev)
                prev = interval
            else:
                prev[1] = max(interval[1], prev[1])

        res.append(prev)
        return res
    
class Solution1:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda i: i[0])
        res = [intervals[0]]

        for interval in intervals[1:]:
            if res[-1][1] < interval[0]:
                res.append(interval)
            else:
                res[-1][1] = max(res[-1][1], interval[1])
        
        return res

File: 0056_mergeIntervals/main.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<std::vector<int>> merge(std::vector<std::vector<int>>& intervals) {
        if(intervals.size() <= 1)
            return intervals;

        sort(intervals.begin(), intervals.end());

        std::vector<std::vector<int>> res;
        res.push_back(intervals[0]);
        for(int i = 1; i < intervals.size(); ++i) {
            // the end of prev interval is smaller than current start, then merge them
            if(res[res.size()-1][1] >= intervals[i][0])
                res[res.size()-1][1] = std::max(res[res.size()-1][1], intervals[i][1]);
            else
                res.push_back(intervals[i]);
        }
        return res;

    }
};